analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/TestAnalysisComparer.cs

```diff
@@ -0,0 +1,132 @@
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.UnitTest.TestFramework
+{
+    public static class TestAnalysisComparer
+    {
+        public static StringBuilder Compare(Diagnostic[] diagnostics, Dictionary<string, IList<IIssueLocation>> expectedIssuesPerFile, string languageVersion)
+        {
+            var summary = new StringBuilder($"Comparing actual issues with expected:\nLanguage version: {languageVersion}.\n");
+            var actualIssues = diagnostics.ToIssueLocations();
+
+            foreach (var fileName in actualIssues.Keys.OrderBy(x => x))
+            {
+                summary.Append('\n').Append(fileName).Append('\n');
+
+                foreach (var actualIssue in actualIssues[fileName].OrderBy(x => x.LineNumber))
+                {
+                    var issue = expectedIssuesPerFile[fileName].FirstOrDefault(x => x.IsPrimary == actualIssue.IsPrimary && x.LineNumber == actualIssue.LineNumber);
+                    summary.Append(Compare(actualIssue, issue));
+                    expectedIssuesPerFile[fileName].Remove(issue);
+                }
+
+                foreach (var expectedIssue in expectedIssuesPerFile[fileName].OrderBy(x => x.LineNumber))
+                {
+                    summary.Append($"Line {expectedIssue.LineNumber}: expected {IssueType(expectedIssue.IsPrimary)} issue {expectedIssue.IssueId ?? "(no id)"} was not raised!\n");
+                }
+            }
+
+            return summary;
+        }
+
+        private static StringBuilder Compare(IIssueLocation actual, IIssueLocation expected)
+        {
+            var summary = new StringBuilder();
+            var prefix = $"Line {actual.LineNumber}";
+            var issueType = IssueType(actual.IsPrimary);
+            var issueId = expected?.IssueId ?? actual.IssueId ?? "(no id)";
+
+            if (expected == null)
+            {
+                summary.Append($"{prefix}: unexpected {issueType} issue {issueId} with message '{actual.Message}'.\n");
+            }
+            else if (expected.Message != null && actual.Message != expected.Message)
+            {
+                summary.Append($"{prefix}: {issueType} issue {issueId} message '{expected.Message}' does not match the actual message '{actual.Message}'.\n");
+            }
+            else if (expected.Start.HasValue && actual.Start != expected.Start)
```

Comment:
The question is what to do if `actual.Start` doesn't have value.

Reply:
I think the behavior is ok now: we will report a problem since we expect something else (the value is different)